enhancements/installer/aws-custom-region-and-endpoints.md

```diff
@@ -0,0 +1,402 @@
+
+#### Configuring the AWS SDK
+
+There are 2 ways to configure the AWS SDK,
+
+1. Use the sdk [resolver][aws-sdk-endpoints-resolver] like Kubernetes AWS [cloud-provider][k8s-aws-cloud-provider-get-resolver] so that the AWS SDK's default resolver is used for AWS APIs and override with the user provided ones when available.
+
+2. Specifically override the AWS config when initializing the clients, for example the terraform-provider-aws's internal [configuration][terraform-provider-aws-config-endpoints]
+
+#### Configuring the internal terraform-provider-aws with service overrides
+
+The Terraform AWS provider doesn't accept random AWS API endpoint [overrides][terraform-aws-provider-api-endpoints], rather endpoint for each service needs to be explicitly specified. So the `tfvars` can pass a map object and the terraform `main.tf` can set each required endpoint override when certain key is present and set in the provide map object.
+
+#### Destroying clusters
+
+The `metadata.json` needs to store the custom endpoints from the `install-config.yaml` so that the users can delete the clusters without any previous state.
+
+### Infrastructure global configuration
+
+Since almost of the cluster operators that communicate to the the AWS API need the use the API endpoints provided by the user, the `Infrastructure` global configuration seems like the best place to store and discover this information.
+
+And in contrast to the current configuration in the `Infrastructure` object like `region` is not editable as day-2 operation, the users should be able to edit the AWS endpoints day-2, therefore this configuration best belongs in the `spec` section.
+
+```go
+// InfrastructureSpec contains settings that apply to the cluster infrastructure.
+type InfrastructureSpec struct {
+  // platformSpec holds configuration specific to the underlying
+  // infrastructure provider.
+  // +optional
+  PlatformSpec *PlatformSpec `json:"platformSpec,omitempty"`
+}
+
+// PlatformSpec holds some configuration to the underlying infrastructure provider
+// of the current cluster. It is supposed that only one of the spec structs is set.
+type PlatformSpec struct {
+  // AWS contains settings specific to the Amazon Web Services infrastructure provider.
+  // +optional
+  AWS *AWSPlatformSpec `json:"aws,omitempty"`
+
+  ...
+}
+```
+
+```go
+// AWSPlatformSpec holds the current status of the Amazon Web Services infrastructure provider.
+type AWSPlatformSpec struct {
+  // ServiceEndpoints list contains custom endpoints which will override default
+  // service endpoint of AWS Services.
+  // There must be only one ServiceEndpoint for a service.
+  // +optional
+  ServiceEndpoints []AWSServiceEndpoint `json:"serviceEndpoints,omitempty"`
+}
+
+// AWSServiceEndpoint store the configuration of a custom url to
+// override existing defaults of AWS Services.
+// Currently Supports - EC2, IAM, ELB, S3 and Route53.
+type AWSServiceEndpoint struct {
```

Comment:
For route53, and resource tagging api,  when you creating the aws client, you need set the api endpoint, you also need set the specific region, like for global you need set to us-east-1, for cn-north-1 and cn-northwest-1, you need set to cn-northwest-1,  do you think it's needed here to add another param `region`? so that match how you create the aws client?

Reply:
the reason you have to set the region is so that the SDK can find the correct endpoint. Since we are already asking the user to give us the endpoint, there is no value for region.